### PR TITLE
Make the ILM and SLM history_index_enabled settings dynamic

### PR DESCRIPTION
**Pull request by joegallo**
Related to https://github.com/elastic/elasticsearch/issues/77466

Makes `indices.lifecycle.history_index_enabled` a dynamic setting, so that it's more useful as an escape hatch in situations where the ILM history writing ends up being a problem (as a dynamic setting, we'd be able to change the value without restarting the cluster).

For symmetry reasons, the same change is also made to `slm.history_index_enabled` although the practical need for a runtime escape hatch there is essentially irrelevant.

### CODE REVIEW

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java [hunk 1]
```diff
@@ -59,12 +58,17 @@ public class ILMHistoryStore implements Closeable {
...
 
-    private final boolean ilmHistoryEnabled;
+    private boolean ilmHistoryEnabled = true;
     private final BulkProcessor processor;
     private final ThreadPool threadPool;
 
-    public ILMHistoryStore(Settings nodeSettings, Client client, ClusterService clusterService, ThreadPool threadPool) {
-        this.ilmHistoryEnabled = LIFECYCLE_HISTORY_INDEX_ENABLED_SETTING.get(nodeSettings);
+    public ILMHistoryStore(Client client, ClusterService clusterService, ThreadPool threadPool) {
+        if (clusterService != null) {
```
Comment (dakrone) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java:66]: I don't think I like having a nullable `ClusterService` for the constructor, is there a way we can remove this check?

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/history/SnapshotHistoryStore.java [hunk 2]
```diff
@@ -38,12 +37,15 @@ public class SnapshotHistoryStore {
...
     private final ClusterService clusterService;
-    private final boolean slmHistoryEnabled;
+    private boolean slmHistoryEnabled = true;
 
-    public SnapshotHistoryStore(Settings nodeSettings, Client client, ClusterService clusterService) {
+    public SnapshotHistoryStore(Client client, ClusterService clusterService) {
         this.client = client;
         this.clusterService = clusterService;
-        slmHistoryEnabled = SLM_HISTORY_INDEX_ENABLED_SETTING.get(nodeSettings);
+        if (clusterService != null) {
```
Comment (dakrone) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/history/SnapshotHistoryStore.java:45]: Same here about cluster service being nullable

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java [hunk 3]
```diff
@@ -59,12 +58,17 @@ public class ILMHistoryStore implements Closeable {
         ).getBytes()
     );
 
-    private final boolean ilmHistoryEnabled;
+    private boolean ilmHistoryEnabled = true;
```
Comment (dakrone) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java:61]: Most of the time we make these `volatile` so they are visible immediately across threads

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/history/SnapshotHistoryStore.java [hunk 4]
```diff
@@ -38,12 +37,15 @@ public class SnapshotHistoryStore {
 
     private final Client client;
     private final ClusterService clusterService;
-    private final boolean slmHistoryEnabled;
+    private boolean slmHistoryEnabled = true;
```
Comment (dakrone) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/history/SnapshotHistoryStore.java:40]: This one should be `volatile` also.

**Review by dakrone — CHANGES_REQUESTED** [review]
I left some comments, I think we shouldn't have a nullable `ClusterService` for these

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java (hunk 3 above)
Comment (joegallo) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java]: Ah, yep, good catch. 71778259957

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/history/SnapshotHistoryStore.java (hunk 4 above)
Comment (joegallo) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/history/SnapshotHistoryStore.java]: Also 71778259957

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java (hunk 1 above)
Comment (joegallo) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java]: How about 33d9a309871b2e352882efff524dc82becb06d69?

**Review by dakrone — APPROVED** [review]
LGTM

**Pull request merged**